src/diff.rs

```diff
@@ -1,103 +1,252 @@
+            ICL: Iterator<Item = T> + Clone,
+            ICR: Iterator<Item = T> + Clone,
+        >(
+            left: ICL,
+            right: ICR,
+            sequence_order: SequenceOrderComparison,
+        ) -> SequenceComparison<T> {
+            match sequence_order {
+                SequenceOrderComparison::Exact => {
+                    Self::strict_order_comparison(left.clone(), right.clone())
+                }
+                SequenceOrderComparison::Relative => {
+                    Self::relative_order_comparison(left.clone(), right.clone())
+                }
+            }
+        }
+
+        pub(self) fn strict_order_comparison<
+            ICL: Iterator<Item = T> + Clone,
+            ICR: Iterator<Item = T> + Clone,
+        >(
+            mut actual_iter: ICL,
+            mut expected_iter: ICR,
+        ) -> SequenceComparison<T> {
+            let mut extra: Vec<T> = vec![];
+            let mut missing: Vec<T> = vec![];
+            let mut order_preserved = true;
+            let move_element = |el: T, source: &mut Vec<T>, target: &mut Vec<T>| {
+                if let Some(idx) = source.iter().position(|e: &T| e.eq(&el)) {
+                    source.remove(idx);
+                } else {
+                    target.push(el);
+                }
+            };
+            loop {
+                match (actual_iter.next(), expected_iter.next()) {
+                    (Some(actual_elem), Some(expect_elem)) => {
+                        if actual_elem.eq(&expect_elem) {
+                            continue;
+                        }
+                        order_preserved = false;
+                        move_element(expect_elem, &mut extra, &mut missing);
+                        move_element(actual_elem, &mut missing, &mut extra);
+                    }
+                    (None, Some(expect_elem)) => {
+                        move_element(expect_elem, &mut extra, &mut missing);
+                    }
+                    (Some(actual_elem), None) => {
+                        move_element(actual_elem, &mut missing, &mut extra);
+                    }
+                    (None, None) => break,
+                }
+            }
+            SequenceComparison {
+                order_preserved,
+                exclusive_left: extra,
+                exclusive_right: missing,
+            }
+        }
+
+        pub(self) fn relative_order_comparison<
+            ICL: Iterator<Item = T> + Clone,
+            ICR: Iterator<Item = T> + Clone,
+        >(
+            mut actual_iter: ICL,
+            mut expected_iter: ICR,
+        ) -> SequenceComparison<T> {
+            let mut actual_value = actual_iter.next();
+            let mut expected_value = expected_iter.next();
+            let mut missing: Vec<T> = vec![];
+            let mut extra: Vec<T> = vec![];
+            loop {
+                if expected_value.is_none() {
+                    extra.extend(actual_iter);
```

Comment:
When `actual_iter=vec![1,2].iter()` and `expected_iter=vec![].iter()`, the first actual element (=1) seems dropped here as `extra` is `[2]`.